core-js/src/definitions.ts

```diff
@@ -0,0 +1,1459 @@
+  addArgument(nameOrArg: string | ArgumentDefinition<DirectiveDefinition>, type?: InputType, defaultValue?: any): ArgumentDefinition<DirectiveDefinition> {
+    let toAdd: ArgumentDefinition<DirectiveDefinition>;
+    if (typeof nameOrArg === 'string') {
+      this.checkUpdate();
+      toAdd = new ArgumentDefinition<DirectiveDefinition>(nameOrArg);
+      toAdd.defaultValue = defaultValue;
+    } else {
+      this.checkUpdate(nameOrArg);
+      toAdd = nameOrArg;
+    }
+    if (this.argument(toAdd.name)) {
+      throw buildError(`Argument ${toAdd.name} already exists on field ${this.name}`);
+    }
+    this._args.set(toAdd.name, toAdd);
+    SchemaElement.prototype['setParent'].call(toAdd, this);
+    if (typeof nameOrArg === 'string') {
+      toAdd.type = type;
+    }
+    return toAdd;
+  }
+
+  get locations(): readonly DirectiveLocationEnum[] {
+    return this._locations;
+  }
+
+  addLocations(...locations: DirectiveLocationEnum[]): DirectiveDefinition {
+    for (const location of locations) {
+      if (!this._locations.includes(location)) {
+        this._locations.push(location);
+      }
+    }
+    return this;
+  }
+
+  addAllLocations(): DirectiveDefinition {
+    return this.addLocations(...Object.values(DirectiveLocation));
+  }
+
+  addAllTypeLocations(): DirectiveDefinition {
+    return this.addLocations('SCALAR', 'OBJECT', 'INTERFACE', 'UNION', 'ENUM', 'INPUT_OBJECT');
+  }
+
+  removeLocations(...locations: DirectiveLocationEnum[]): DirectiveDefinition {
+    for (const location of locations) {
+      const index = this._locations.indexOf(location);
+      if (index >= 0) {
+        this._locations.splice(index, 1);
+      }
+    }
+    return this;
+  }
+
+  private addReferencer(referencer: Directive<TApplicationArgs>) {
+    assert(referencer, 'Referencer should exists');
+    this._referencers.add(referencer);
+  }
+
+  protected removeTypeReference(type: NamedType) {
+    assert(false, `Directive definition ${this} can't reference other types (it's arguments can); shouldn't be asked to remove reference to ${type}`);
+  }
+
+  remove(): Directive[] {
+    if (!this._parent) {
+      return [];
+    }
+    Schema.prototype['removeDirectiveInternal'].call(this._parent, this);
+    this._parent = undefined;
+    for (const directive of this._appliedDirectives) {
+      directive.remove();
+    }
+    for (const arg of this._args.values()) {
+      arg.remove();
+    }
+    // Note that directive applications don't link directly to their definitions. Instead, we fetch
+    // their definition from the schema when requested. So we don't have to do anything on the referencers
+    // other than return them.
+    const toReturn = [... this._referencers];
+    this._referencers.clear();
+    return toReturn;
+  }
+
+  toString(): string {
+    return this.name;
+  }
+}
+
+// TODO: How do we deal with default values? It feels like it would make some sense to have `argument('x')` return the default
```

Comment:
That's a good question. I would say it makes sense for `arguments` (or perhaps `argumentValues`) to return the resolved argument values that take default values into account (that's also consistent with `getArgumentValues()`/`getDirectiveValues()` in `graphql-js`. But I don't think default values should be included in the internal representation, because you want to differentiate between explicitly providing a value and relying on the default value. That also means changing the default value in the directive definition should affect `argumentValues`, making that a function that depends on the directive definition (again, like `getArgumentValues`).

Reply:
Also see https://github.com/graphql/graphql-js/pull/3049 for some background on default value semantics.

Reply:
I definitively agree about not storing the default values internally, that wasn't my intention.

I'll note thought that you're suggestion about having optional `TArgs` for directives make this a tad less obvious. Because with that, it made sense to remove the `Directive.argument(name: string): any` method, and always rely on `get Directive.arguments(): TArgs` (I mean, what's the point of doing `directive.argument('foo')` and potentially losing type checking, when you can do `directive.arguments.foo` and not lose type checking).

But it's a tad less clear how to handle defaults in that case (API wise I mean).

I can think of a few options:
1. stop having `Directive.arguments` be a getter, and be `Directive.arguments(withDefaultValues: boolean = false): TArgs`. If you pass `withDefaultValues = true`, we'd copy the underlying args object and set undefined keys to their defaults if necessary.
2. keep `Directive.arguments` a getter, have it always include default values but have an additional `isDefaultValue(name: string)` for if you want to know if a particular value was provided.
3. keep `Directive.arguments` a getter, have it not include default values and have a separate method `argumentOrDefault(na me: string): any` that can alternative give you values with defaults. I like this less because you don't get type-checked access if you also want defaults.

Overall, leaning towards option 1.